infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/provisioner/labels/SinglePortLabelsProvisioner.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.docker.provisioner.labels;
+
+import static java.lang.String.format;
+
+import java.util.HashMap;
+import java.util.Map;
+import javax.inject.Inject;
+import javax.inject.Named;
+import org.eclipse.che.api.core.model.workspace.config.ServerConfig;
+import org.eclipse.che.api.core.model.workspace.runtime.RuntimeIdentity;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.api.workspace.server.spi.environment.InternalMachineConfig;
+import org.eclipse.che.commons.annotation.Nullable;
+import org.eclipse.che.workspace.infrastructure.docker.model.DockerContainerConfig;
+import org.eclipse.che.workspace.infrastructure.docker.model.DockerEnvironment;
+import org.eclipse.che.workspace.infrastructure.docker.provisioner.ConfigurationProvisioner;
+import org.eclipse.che.workspace.infrastructure.docker.server.mapping.SinglePortHostnameBuilder;
+
+/**
+ * Sets necessary container labels for the single-port proxy (Traefik).
+ *
+ * @author Max Shaposhnik (mshaposh@redhat.com)
+ */
+public class SinglePortLabelsProvisioner implements ConfigurationProvisioner {
+
+  private final SinglePortHostnameBuilder hostnameBuilder;
+
+  @Inject
+  public SinglePortLabelsProvisioner(
+      @Nullable @Named("che.docker.ip.external") String externalIpOfContainers,
+      @Named("che.docker.ip") String internalIpOfContainers) {
+    this.hostnameBuilder =
+        new SinglePortHostnameBuilder(externalIpOfContainers, internalIpOfContainers);
+  }
+
+  @Override
+  public void provision(DockerEnvironment internalEnv, RuntimeIdentity identity)
+      throws InfrastructureException {
+    for (Map.Entry<String, InternalMachineConfig> machineEntry :
+        internalEnv.getMachines().entrySet()) {
+      final String machineName = machineEntry.getKey();
+      Map<String, String> containerLabels = new HashMap<>();
+      for (Map.Entry<String, ServerConfig> serverEntry :
+          machineEntry.getValue().getServers().entrySet()) {
+        final String serverName = serverEntry.getKey().replace('/', '-');
```

Comment:
I believe that server name can contain other symbols that are not allowed in the DNS entry. Can you check whether such a use case breaks this code?